RFC: Vector library (#34)
luau-lang/rfcs: docs/vector-library.md

# Vector library

## Summary

Implement a standard library that provides functionality for the vector type.

## Motivation

Currently, vectors are a primitive type implemented internally. All of the heavy work to implement vectors is done, but there is no runtime-agnostic way to use vectors in Luau code. Individual Luau runtimes, such as [Lune](https://github.com/lune-org/lune) or Roblox, must provide their own distinct libraries to work with native vectors which may be inconsistent. In cross-runtime code, this results in performance drawbacks & difficulty utilizing native vectors.

## Design

The default metatable for vectors should now be the `vector` library. While buffers and coroutines do not have this, vectors likely should for ergonomics.

### Library functions & constants

This RFC proposes the following basic functions & constants as a starting point, but as with all builtin libraries, future RFCs can propose additional functions.

---

> `vector(x: number?, y: number?, z: number?)`

Reviewer: ```suggestion
`vector(x: number, y: number, z: number)`
```

It's best to encourage practice of hitting the fastest constructor we have available.

It will also avoid confusion of what vector(1) is.
While `Vector3.new(1)` is [1, 0, 0] on Roblox, for people coming from GLSL where `vec3(1)` is [1, 1, 1], it might be unintuitive.
Even if supported by runtime, the typing can be more strictly defined.

Creates a vector with 3 components: x, y, z. If the feature flag for wide vectors is enabled, a fourth argument `w: number?` will be introduced.

Due to the common usage of vectors, vector creation should be ergonomic. Therefore, it is probably worth breaking the `create()` naming standard.

`vector.magnitude(vecA: vector): number`

Calculates the magnitude of a given vector.

`vector.normalized(vec: vector): vector`

Returns the normalized version (aka unit vector) of a given vector. If a zero vector is passed, return zero. For further clarification, `vector.normalized(vector(0, 0, 0))` should return zero.

`vector.cross(vecA: vector, vecB: vector): vector`

Returns the cross product of two vectors. If 4-wide vectors are enabled, this function will ignore the fourth component, and return the 3-dimensional cross product.

> `vector.dot(vecA: vector, vecB: vector): vector`

Reviewer: dot product results in a number.

Returns the dot product of two vectors.

`vector.floor(vec: vector): vector`

Equivalent of `math.floor` for vectors.

`vector.ceil(vec: vector): vector`

Equivalent of `math.ceil` for vectors.

`vector.angle(vecA: vector, vecB: vector): vector`

> Returns the angle between two vectors.

Reviewer: In what units?

# Vector library

## Summary

Implement a standard library that provides functionality for the vector type.

## Motivation

Currently, vectors are a primitive type implemented internally. All of the heavy work to implement vectors is done, but there is no runtime-agnostic way to use vectors in Luau code. Individual Luau runtimes, such as [Lune](https://github.com/lune-org/lune) or Roblox, must provide their own distinct libraries to work with native vectors which may be inconsistent. In cross-runtime code, this results in performance drawbacks & difficulty utilizing native vectors.

## Design

> The default metatable for vectors should now be the `vector` library. While buffers and coroutines do not have this, vectors likely should for ergonomics.

Reviewer: ```suggestion
The default metatable for vectors should now be the `vector` library. While buffers and coroutines do not have this, vectors likely should for ergonomics.

Luau VM can be configured in two different vector value modes.
Default configuration uses vectors with 3 components (`xyz`) and if the `LUA_VECTOR_SIZE` configuration option is set to 4, vector values get an additional forth `w` component.
This mode will be referred to as '4-wide mode' in this proposal.
```

### Library functions & constants

This RFC proposes the following basic functions and constants.

> `vector.create(x: number, y: number, z: number): vector`
> 
> Creates a new vector with the given components. When in 4-wide mode takes a fourth `w` argument. None of the arguments are optional.

Reviewer: Constructor is something that already has a defined behavior in both compiler and runtime.
This change will not be backwards compatible for 4-wide vector mode.

For 4-wide vectors I propose a different description:

```suggestion
`vector.create(x: number, y: number, z: number): vector`

Creates a new vector with the given components. In 4-wide mode, the forth `w` component value is set to 0.0.

`vector.create(x: number, y: number, z: number, w: number): vector`

Additional constructor available in 4-wide mode. Creates a new vector with the given components.
```

I think this matches the compiler behavior where `vector.create(1, 2, 3, nil)` will not be recognised as a constant and the fast-call behavior where again, `vector.create(1, 2, 3, nil)` will exit the fast-call into a library call fallback. 
`vector.create(1, 2, 3)` is allowed in constant folding and fast-calls in 4-wide mode today.

While we do define the fallback library function in Conformance.test.cpp to treat explicit `nil` as 0.0, we can change the method to use the new library one, since it's just a unit test.

`vector.magnitude(vec: vector): number`

Calculates the magnitude of a given vector. When in 4-wide mode, this includes the fourth component.

`vector.normalize(vec: vector): vector`

Returns the normalized version (aka unit vector) of a given vector. When in 4-wide mode, this includes the fourth component.

`vector.cross(vec1: vector, vec2: vector): vector`

Returns the cross product of two vectors. If 4-wide vectors are enabled, this function will ignore the fourth component and return the 3-dimensional cross product.

`vector.dot(vec1: vector, vec2: vector): number`

Returns the dot product of two vectors. Uses all components, even in 4-wide mode.

`vector.angle(vec1: vector, vec2: vector, axis: vector?): number`

Returns the angle between two vectors in radians. The axis, if specified, is used to determine the sign of the angle. If 4-wide vectors are enabled, this function will ignore the fourth component and return the 3-dimensional angle.

`vector.floor(vec: vector): vector`

Applies `math.floor` component-wise for each vector. Equivalent of: `vector(math.floor(vec.x), etc)`.

`vector.ceil(vec: vector): vector`

> Applies `math.ceil` component-wise for each vector. Equivalent of: `vector(math.ceil(vec.x), etc)`.

Reviewer: ```suggestion
Applies `math.ceil` component-wise for each vector. Equivalent of: `vector.create(math.ceil(vec.x), etc)`.
```

`vector.abs(vec: vector): vector`

> Applies `math.abs` component-wise for each vector. Equivalent of: `vector(math.abs(vec.x), etc)`.

Reviewer: ```suggestion
Applies `math.abs` component-wise for each vector. Equivalent of: `vector.create(math.abs(vec.x), etc)`.
```

`vector.sign(vec: vector): vector`

> Applies `math.sign` component-wise for each vector. Equivalent of: `vector(math.sign(vec.x), etc)`.

Reviewer: ```suggestion
Applies `math.sign` component-wise for each vector. Equivalent of: `vector.create(math.sign(vec.x), etc)`.
```

`vector.clamp(vec: vector, min: vector, max: vector): vector`

> Applies `math.clamp` component-wise for each vector. Equivalent of: `vector(math.clamp(vec.x, min.x, max.x), etc)`.

Reviewer: ```suggestion
Applies `math.clamp` component-wise for each vector. Equivalent of: `vector.create(math.clamp(vec.x, min.x, max.x), etc)`.
```

`vector.max(...: vector): vector`

> Applies `math.max` component-wise for each vector. Equivalent of: `vector(math.max((...).x), etc)`.

Reviewer: ```suggestion
Applies `math.max` component-wise for each vector. Equivalent of: `vector.create(math.max((...).x), etc)`.
```

`vector.min(...: vector): vector`

> Applies `math.min` component-wise for each vector. Equivalent of: `vector(math.min((...).x), etc)`.

Reviewer: ```suggestion
Applies `math.min` component-wise for each vector. Equivalent of: `vector.create(math.min((...).x), etc)`.
```

## Motivation

Currently, vectors are a primitive type implemented internally. All of the heavy work to implement vectors is done, but there is no runtime-agnostic way to use vectors in Luau code. Individual Luau runtimes, such as [Lune](https://github.com/lune-org/lune) or Roblox, must provide their own distinct libraries to work with native vectors which may be inconsistent. In cross-runtime code, this results in performance drawbacks & difficulty utilizing native vectors.

## Design

The default metatable for vectors should now be the `vector` library. While buffers and coroutines do not have this, vectors likely should for ergonomics.

### Library functions & constants

This RFC proposes the following basic functions & constants as a starting point, but as with all builtin libraries, future RFCs can propose additional functions.

---

`vector(x: number, y: number, z: number)`

Creates a vector with 3 components: x, y, z. If the feature flag for wide vectors is enabled, a fourth argument `w: number?` will be introduced.

Due to the common usage of vectors, vector creation should be ergonomic. Therefore, it is probably worth breaking the `create()` naming standard.

> `vector.magnitude(vecA: vector): number`

Reviewer: ```suggestion
`vector.magnitude(vec: vector): number`
```

Calculates the magnitude of a given vector.

`vector.normalized(vec: vector): vector`

> Returns the normalized version (aka unit vector) of a given vector. If a zero vector is passed, return zero. For further clarification, `vector.normalized(vector(0, 0, 0))` should return zero.

Reviewer: For compatibility with Roblox runtime and for code size and performance (including branchless native codegen lowering) I suggest removing requirements for zero length vector detection.

Author: That makes sense.

`vector.cross(vecA: vector, vecB: vector): vector`

Returns the cross product of two vectors. If 4-wide vectors are enabled, this function will ignore the fourth component, and return the 3-dimensional cross product.

`vector.dot(vecA: vector, vecB: vector): number`

> Returns the dot product of two vectors.

Reviewer: Can we note how it behaves for 4-wide vectors?

Reviewer 2: I'm unsure of the behavior it should have - I would simply use only the first three components even when compiled in four component mode. The other options are to error or compute the dot product of vectors with four components, which a quick google returns little results for.

Reviewer 3: I don't think having it only work on the first 3 components is correct.
If we were to follow the definition of the dot product between 2 vectors
$$a \cdot b = \sum_{i=1}^n a_i b_i$$
we can say that the dot product for four component vectors would be:
$$\vec{a} \cdot \vec{b} = a_1 b_1 + a_2 b_2 + a_3 b_3 + a_4 b_4 $$

Reviewer 2: resolved

`vector.floor(vec: vector): vector`

Equivalent of `math.floor` for vectors.

`vector.ceil(vec: vector): vector`

Equivalent of `math.ceil` for vectors.

> `vector.angle(vecA: vector, vecB: vector): number`

Author: No.

Returns the angle between two vectors in radians.

`vector.abs(vec: vector): vector`

Applies math.abs component-wise for each vector. Equivalent of: `vector(math.abs(vec.x), ect)`.

`vector.sign(vec: vector): vector`

Applies `math.sign` component-wise for each vector. Equivalent of: `vector(math.sign(vec.x), ect)`.

`vector.clamp(vec: vector, min: vector, max: vector): vector`

Applies `math.clamp` component-wise for each vector. Equivalent of: `vector(math.clamp(vec.x, min.x, max.x), ect)`.

`vector.max(...: vector): vector`

Applies `math.max` component-wise for each vector. Equivalent of: `vector(math.max((...).x), ect)`.

`vector.min(...: vector): vector`

Applies `math.min` component-wise for each vector. Equivalent of: `vector(math.min((...).x), ect)`.

---

`vector.zero`

Vector where `x=0, y=0, z=0, w?=0`.

`vector.one`

Vector where `x=1, y=1, z=1, w?=1`.

---

### Arithmetic operations

Primitive operators for vectors are already implemented, so this RFC doesn't concern vector arithmetic.

### Native codegen

> In the future, vectors will have special treatment in native codegen. This creates an expectation that the vector library will also have special treatment in codegen, but it isn't clear what that will look like, or if the performance benefits make sense.

Reviewer: This can be removed as vectors already have special treatment in codegen.

Generally, library proposals shouldn't have to think about fastcalls/lowering hooks as those are implementation details which are implemented without a proposal being required.
Whether or not existing custom lowering hooks from Roblox will be included (or written from scratch) doesn't have to be specified.

Author: Sounds good

### Typechecking

A new primitive `vector` type should be added to the type system.

## Drawbacks

As per all additional globals, this creates a new global `vector`. This is a commonly used variable name, and many style guides will advise to avoid using `vector` as a variable name due to the global being added.

Introducing a vector library means introducing more globals. Due to the prescence of vectors in performance-intensive code, it can be assumed that the built-in functions will have fastcalls. This means more fastcall slots will be used.

Existing code won't get any improvements from a built-in library. Developers will have to switch over to the built-in library in order to gain any benefit.

## Alternatives

Do nothing; vectors have an internal constructor, which lets the runtime implement vectors, and subsequently a library for vector math.

### Alternative implementations

> A more standard alternative to `vector(...)` could be something like `vector.create` or `vector.new`. This follows the standard set by `buffer.create`, `coroutine.create`, and `table.create`, and doesn't involve calling a table, which requires a metatable with `__call` to be set on the library. This breaks the standard previously set by Lua, as there is currently no library with a metatable set by default.

Reviewer: string library sets the metatable of string values by default, so there is precedent (`luaopen_string`).

Author: To clarify, I mean there is no library which has an implicit `__call` metamethod on it. There's no luau library that lets you do, for example, `string()`.

Reviewer: That's not what the RFC text says. It still says "there is currently no library with a metatable set by default" and that's not true. If it said that there's no library with a `__call` metamethod set on a type, that would be true.

Reviewer 2: resolved

`vector.floor(vec: vector): vector`

Applies `math.floor` component-wise for each vector. Equivalent of: `vector.create(math.floor(vec.x), etc)`.

`vector.ceil(vec: vector): vector`

Applies `math.ceil` component-wise for each vector. Equivalent of: `vector.create(math.ceil(vec.x), etc)`.

`vector.abs(vec: vector): vector`

Applies `math.abs` component-wise for each vector. Equivalent of: `vector.create(math.abs(vec.x), etc)`.

`vector.sign(vec: vector): vector`

Applies `math.sign` component-wise for each vector. Equivalent of: `vector.create(math.sign(vec.x), etc)`.

`vector.clamp(vec: vector, min: vector, max: vector): vector`

Applies `math.clamp` component-wise for each vector. Equivalent of: `vector.create(math.clamp(vec.x, min.x, max.x), etc)`.

`vector.max(...: vector): vector`

Applies `math.max` component-wise for each vector. Equivalent of: `vector.create(math.max((...).x), etc)`.

`vector.min(...: vector): vector`

Applies `math.min` component-wise for each vector. Equivalent of: `vector.create(math.min((...).x), etc)`.

---

`vector.zero`

Vector where `x=0, y=0, z=0, w?=0`.

`vector.one`

Vector where `x=1, y=1, z=1, w?=1`.

---

### Arithmetic operations

> Primitive operators for vectors are already implemented, so this RFC doesn't concern vector arithmetic.

Reviewer: ```suggestion
Primitive operators for vectors are already implemented, so this RFC doesn't concern vector arithmetic.

### Component access

Component access is already supported in the VM using the fields 'x', 'y', 'z', 'X', 'Y' and 'Z' (plus 'w' and 'W' in 4-wide mode).
Note that writes to a single component are not supported as the vector value is immutable.
This RFC doesn't define any changes to this behavior.

```

`vector.abs(vec: vector): vector`

Applies math.abs component-wise for each vector. Equivalent of: `vector(math.abs(vec.x), ect)`.

`vector.sign(vec: vector): vector`

Applies `math.sign` component-wise for each vector. Equivalent of: `vector(math.sign(vec.x), ect)`.

`vector.clamp(vec: vector, min: vector, max: vector): vector`

Applies `math.clamp` component-wise for each vector. Equivalent of: `vector(math.clamp(vec.x, min.x, max.x), ect)`.

`vector.max(...: vector): vector`

Applies `math.max` component-wise for each vector. Equivalent of: `vector(math.max((...).x), ect)`.

`vector.min(...: vector): vector`

Applies `math.min` component-wise for each vector. Equivalent of: `vector(math.min((...).x), ect)`.

---

`vector.zero`

Vector where `x=0, y=0, z=0, w?=0`.

`vector.one`

Vector where `x=1, y=1, z=1, w?=1`.

---

### Buffer Library

`buffer.writevector(b: buffer, offset: number, vec: vector)`

Writes the vector into the buffer at the offset. Each component will be written as a f32. In four component mode, it will write all four components and write 16 bytes. In three component mode, it will write 12 bytes.

`buffer.readvector(b: buffer, offset: number): vector`

> Reads a vector from the buffer at the offset. In four component mode it will read 16 bytes and in three component mode it will read 12 bytes.

Reviewer: This might not be the path for vector storage in buffers.
I would suggest taking a look at the discussion we had on this topic: https://github.com/luau-lang/luau/discussions/1296

I think a good path forward would be `readvector2`, `readvector3` and `readvector4` with similar methods to write.
I don't think f32xN is a good choice because it implies multiple value read/write unrelated to the vector type.

I do like the idea of having 2-component access as an option.

Reviewer 2: functions removed and your idea has been placed into future work

Reviewer: I can create a buffer update later when I go over pending buffer RFCs as well.

# Vector library

## Summary

Implement a standard library that provides functionality for the vector type.

## Motivation

Currently, vectors are a primitive type implemented internally. All of the heavy work to implement vectors is done, but there is no runtime-agnostic way to use vectors in Luau code. Individual Luau runtimes, such as [Lune](https://github.com/lune-org/lune) or Roblox, must provide their own distinct libraries to work with native vectors which may be inconsistent. In cross-runtime code, this results in performance drawbacks & difficulty utilizing native vectors.

## Design

The default metatable for vectors should now be the `vector` library. While buffers and coroutines do not have this, vectors likely should for ergonomics.

Luau VM can be configured in two different vector value modes.

> Default configuration uses vectors with 3 components (`xyz`) and if the `LUA_VECTOR_SIZE` configuration option is set to 4, vector values get an additional forth `w` component.

Reviewer 4: Do we support other values for `LUA_VECTOR_SIZE` other than 3 and 4? If so, how will these other values be supported (or not)?

Reviewer: We do not support any other configurations.

> Default configuration uses vectors with 3 components (`xyz`) and if the `LUA_VECTOR_SIZE` configuration option is set to 4, vector values get an additional forth `w` component.

Reviewer 5: ```suggestion
Default configuration uses vectors with 3 components (`xyz`) and if the `LUA_VECTOR_SIZE` configuration option is set to 4, vector values get an additional fourth `w` component.
```

This mode will be referred to as '4-wide mode' in this proposal.

### Library functions & constants

This RFC proposes the following basic functions and constants.

`vector.create(x: number, y: number, z: number): vector`

> Creates a new vector with the given components. In 4-wide mode, the forth `w` component value is set to 0.0.

Reviewer 5: ```suggestion
Creates a new vector with the given components. In 4-wide mode, the fourth `w` component value is set to 0.0.
```

### Component access

Component access is already supported in the VM using the fields 'x', 'y', 'z', 'X', 'Y' and 'Z' (plus 'w' and 'W' in 4-wide mode).
Note that writes to a single component are not supported as the vector value is immutable.
This RFC doesn't define any changes to this behavior.

### Compiler options

Currently, there are 2 compiler options relating to vectors, `vectorLib` & `vectorCtor`. This poses an interesting problem: a builtin library would remove the _requirement_ for such compiler options, however these options still need to be supported. The intuitive solution to this is to leave both compiler options working and maintained, but provide the built-in vector library by default, allowing two vector libraries and two vector constructors.

### Typechecking

A new primitive `vector` type should be added to the type system.

## Future Work

A better vector constructor with new syntax, such as `<x, y, z>` or `(x, y, z)` or `|x, y, z|` or `vector(x, y, z)` or `[x, y, z]` could be implemented. This would make code that does stuff with vectors much less verbose.

> The buffer library could include functions to work with vectors, such as `buffer.writevector2`, `buffer.writevector3`, and in 4-wide mode, `buffer.writevector4`. The corrosponding read functions would of course be included.

Reviewer 4: ```suggestion
The buffer library could include functions to work with vectors, such as `buffer.writevector2`, `buffer.writevector3`, and in 4-wide mode, `buffer.writevector4`. The corresponding read functions would of course be included.
```

## Drawbacks

As per all additional globals, this creates a new global `vector`. This is a commonly used variable name, and many style guides will advise to avoid using `vector` as a variable name due to the global being added.

> Introducing a vector library means introducing more globals. Due to the prescence of vectors in performance-intensive code, it can be assumed that the built-in functions will have fastcalls. This means more fastcall slots will be used.

Reviewer: ```suggestion
Introducing a vector library means introducing more globals. Due to the presence of vectors in performance-intensive code, it can be assumed that the built-in functions will have fastcalls. This means more fastcall slots will be used.
```
